Fixing jax integration for latest jax versions

taylorbell57 opened the pull request:
These proposed edits were all tested with jax-0.5.3

Took a lot of back and forth with ChatGPT and debugging/testing, but now harmonica mostly seems to work well for me with jax and numpy fitters.

There's still some issues with a/b values that are ~0 or <0 when using the jax autodiff version that I'm still working to resolve.

This PR resolves #16

taylorbell57 committed:
Moving bindings to avoid circular imports

taylorbell57 committed:
Functional but not very pretty fix to custom_primitives.py for updated jax version

taylorbell57 committed:
Fixing jax code so that it now runs properly and fits properly, and all tests pass

taylorbell57 committed:
Incrementing version

taylorbell57 committed:
Initial attempts at handling r values at, near, or below 0.

taylorbell57 commented:
FYI, I think this PR still doesn't work - it worked in some ways, but the autodiff didn't seem to work for fitting for reasons I hadn't worked out yet

DavoGrant commented:
@taylorbell57 totally awesome -- thanks loads for putting in the time to keep jax working with harmonica. Great to hear you have been adding it into Eureka! too. Merging this pr now and I will make a new release for 0.2.1, as proposed.

taylorbell57 commented:
Note my comment above though - the code in this PR draft makes some progress on getting things working, but it still doesn't work properly

DavoGrant commented:
Okay, I'll have to look into that. It has passed the tests, at least! Perhaps we need a test fit added too.

kevin218 commented:
Once this PR is merged, it would be great to have new pre-built wheels for python 3.11 - 3.13.

taylorbell57 committed:
Fixing issues caused turing eclipse phase

taylorbell57 committed:
Adding temporary visual test for quad LD

taylorbell57 committed:
Adding more fixes to better handle r values near 0

taylorbell57 committed:
Add tests for stability near zero ripple coefficients. Add more debugging logs

taylorbell57 commented:
Well I made more progress today, but there is still a numerical issue with spikes during the start of ingress and the end of egress for quadratic LD with this draft of the code. I'll circle back to this in a while

taylorbell57 committed:
Removing counter-productive safety guards in light_curve/gradients.cpp

taylorbell57 committed:
Reverting some changes to fluxes.cpp

taylorbell57 committed:
flake8

taylorbell57 committed:
Add assertions to tests to make sure fluxes are [0,1]

taylorbell57 committed:
Removing temporary test_quadLD.py

taylorbell57 committed:
Modify setup.py to use find_packages for package discovery

taylorbell57 committed:
Fix deprecated ndarray-to-scalar assignment in light_curve loop

Avoid assigning a 1-element NumPy array to a scalar slot in `self.fs[i]`
by extracting the scalar with `f[0]`. This prevents a DeprecationWarning
in NumPy 1.25+ and ensures future compatibility.

taylorbell57 committed:
Cleaning up some old, no-longer-needed edits

taylorbell57 commented:
Made more progress last night and now all the tests pass, but I'm now encountering some strange behaviours when I try to actually use the harmonica.jax code. It's possible it's a Eureka! bug instead of a harmonica bug, but I'll have to investigate more.

taylorbell57 committed:
Adding init files to make sure all folders are discovered automatically

taylorbell57 committed:
Making sure all needed files are included when building

taylorbell57 committed:
Add support for flipping transit if rp[0]<0 for HarmonicaTransit

taylorbell57 committed:
Modify _harmonica_transit_common for improved clarity and robustness in ripple coefficient handling

taylorbell57 committed:
Fix flipping logic for transit in HarmonicaTransit to properly handle multi-dimensional input

hrwakeford commented:
@taylorbell57 where do you feel you are with this pull request? What steps are remaining?

taylorbell57 commented:
It has been a while since I worked on this, so my memory is a bit foggy. From what I recall I had initially encountered some weird behaviours near Rp/Rs ~ 0; I think I might have largely worked that out (by making Rp/Rs < 0 produce an upside-down transit signal), but I could be wrong. The other issue I was having was that Harmonica.jax was incredibly slow when I was testing it within Eureka!, but so were basically all of the other jax methods I was testing within Eureka! (because I was using object-oriented programming which breaks jax jitting/tracing). Now that I have a mostly functioning jax-version of Eureka!, I'll hopefully be able to test this PR fairly soon. But that said, if you @hrwakeford or a student also has a bit of time to test out this PR, that'd be super helpful

taylorbell57 commented:
@hrwakeford, I've just managed to test this on a dataset with Eureka!, and I was able to do optimization and sampling successfully (though I just used a circular planet for my test). The NUTS sampling with a circular planet from harmonica was quite a bit slower (≥2x) than for a jaxoplanet transit model though, so it is possible/probable that there are still some improvements that can be made but at least it doesn't outright crash. My fit also didn't get anywhere near Rp/Rs~0, so I don't know if the code still doesn't handle that singularity nicely

taylorbell57 commented:
I also just ran a successful test with a1 and b1 as well!

hrwakeford commented:
@taylorbell57  I am assigning @cefairman & @cathal-maguire as reviewers for this. Both please run some tests on the changes made by Taylor and make sure it does not impact current functionality of the code

hrwakeford commented:
Once this pull request has been tested and approved I will merge it with the current module and then produce a new release of the package so that older versions without this compatibility/set-up can be run if necessary

hrwakeford commented:
@taylorbell57 @kevin218 Please note that we are locking this version of harmonica in place as linked to the paper and original format of the code. 

If possible could you please issue this pull request on the forked version in the ExoTiC collaboration https://github.com/Exo-TiC/harmonica as this will be the version of the code which will be maintained and released with up-to-date modifications. 

I will remove @cathal-maguire and @cefairman as reviewers here and then I will close the PR here once it is issued on the other repository. 

The forked version on ExoTiC will now be where future branches are made and any PRs issued here will be closed directing people to the fork.
